beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -0,0 +1,226 @@
+                        .collect::<Vec<_>>();
+                    let data_columns_result = DataColumnSidecar::build_sidecars(
+                        &blob_refs,
+                        &block_cloned,
+                        &kzg,
+                        &spec,
+                    );
+
+                    let all_data_columns = match data_columns_result {
+                        Ok(d) => d,
+                        Err(e) => {
+                            error!(logger, "Failed to build data column sidecars from blobs"; "error" => ?e);
+                            return;
+                        }
+                    };
+
+                    // Check indices from cache before sending the columns, to make sure we don't
+                    // publish components already seen on gossip.
+                    let all_data_columns_iter = all_data_columns.clone().into_iter();
+                    let data_columns_to_publish = match chain_cloned.data_availability_checker.imported_custody_column_indexes(&block_root) {
+                        None => Either::Left(all_data_columns_iter),
+                        Some(imported_columns_indices) => Either::Right(all_data_columns_iter.filter(move |d| !imported_columns_indices.contains(&d.index()))),
+                    }.collect::<Vec<_>>();
+
+                    if let Err(e) = data_columns_sender.try_send(all_data_columns) {
+                        error!(logger, "Failed to send computed data columns"; "error" => ?e);
+                    };
+
+                    let is_supernode =
+                        chain_cloned.data_availability_checker.get_custody_columns_count() == spec.number_of_columns;
+                    if is_supernode && !data_columns_to_publish.is_empty() {
+                        publish_fn(BlobsOrDataColumns::DataColumns(data_columns_to_publish));
+                    }
+                },
+                "compute_data_columns",
+            )
+            .ok_or(BeaconChainError::RuntimeShutdown)?;
+    } else {
+        let all_blobs = fixed_blob_sidecar_list.clone();
+        let all_blobs_iter = all_blobs.into_iter().flat_map(|b| b.clone());
+
+        let blobs_to_publish = match chain
+            .data_availability_checker
+            .imported_blob_indexes(&block_root)
+        {
+            None => Either::Left(all_blobs_iter),
+            Some(imported_blob_indices) => Either::Right(
+                all_blobs_iter.filter(move |b| !imported_blob_indices.contains(&b.index())),
+            ),
+        };
+
+        publish_fn(BlobsOrDataColumns::Blobs(
+            blobs_to_publish.collect::<Vec<_>>(),
+        ));
+    };
+
+    debug!(
+        chain.log,
+        "Blobs from EL - start processing";
+        "num_blobs" => num_blobs,
+    );
+
+    chain
+        .process_engine_blobs(
+            block.slot(),
+            block_root,
+            fixed_blob_sidecar_list.clone(),
+            peer_das_enabled.then_some(data_columns_receiver),
```

Comment:
There's a mismatch here, `data_columns_sender` is only used if `peer_das_enabled && all_blobs_fetched` not only `peer_das_enabled`

Reply:
You can move `let (data_columns_sender, data_columns_receiver) = tokio::sync::mpsc::channel(1);` to inside the if block and return the receiver from the if else

Reply:
Good catch, fixed